diffrax/_solver/srk.py

```diff
@@ -0,0 +1,649 @@
+import abc
+from dataclasses import dataclass
+from typing import ClassVar, Optional, TYPE_CHECKING, Union
+from typing_extensions import TypeAlias
+
+import equinox as eqx
+import equinox.internal as eqxi
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+import numpy as np
+from equinox.internal import ω
+from jaxtyping import Array, Float, PyTree
+
+from .._brownian import AbstractBrownianPath
+from .._custom_types import (
+    BoolScalarLike,
+    DenseInfo,
+    FloatScalarLike,
+    IntScalarLike,
+    LevyVal,
+    RealScalarLike,
+    VF,
+    Y,
+)
+from .._local_interpolation import LocalLinearInterpolation
+from .._solution import RESULTS
+from .._term import AbstractTerm, MultiTerm, ODETerm
+from .base import AbstractSolver
+
+
+if TYPE_CHECKING:
+    from typing import ClassVar as AbstractClassVar
+else:
+    from equinox import AbstractClassVar
+
+_ErrorEstimate: TypeAlias = Optional[Y]
+_SolverState: TypeAlias = None
+_CarryType: TypeAlias = tuple[PyTree[Array], PyTree[Array], PyTree[Array]]
+
+
+@dataclass(frozen=True)
+class AbstractNoiseCoefficients:
+    """Abstract base class for noise coefficients."""
+
+    a: Union[Float[np.ndarray, " s"], tuple[np.ndarray, ...]]
+    b: Union[FloatScalarLike, Float[np.ndarray, " s"]]
```

Comment:
I don't think we need any annotations here at all. Are we ever doing:
```python
def foo(x: AbstractNoiseCoefficients):
    bar(x.a)
```
i.e. relying on the presence of these variables with these (quite broad) types?

Reply:
We do need the presence of these variables, e.g. in `a_w = jnp.asarray(self.tableau.cfs_w.a, dtype=dtype)`. We don't rely on the types, but I'm not sure how to declare an attribute without a type.

Reply:
@patrick-kidger I'm still not sure how to get rid of these annotations in this piece of code:

https://github.com/andyElking/diffrax_STLA/blob/1087d825fa5bcf49017ead81628a71b69e42566b/diffrax/_solver/srk.py#L46-L50

Reply:
Also why does this not turn into a code snippet??

Reply:
I belive only links from the same repository will turn into code snippets.

Regarding what the code here now looks like: I think I'd encourgae you to "prefer composition over inheritance" -- this is one of those old adages in compsci. Right now you have a single `AbstractBrownianCoeffs` with many different variations. However I think you could do something like this instead:
```python
class AbstractLevyAreaCoeffs:
    ...

class SpaceTimeLevyAreaCoeffs(AbstractLevyAreaCoeffs):
    ...

class SpaceTimeTimeLevyAreaCoeffs(AbstractLevyAreaCoeffs):
    ...

class AbstractBrownianCoeffs:
    ...

class AdditiveBrownianCoeffs(AbstractBrownianCoeffs):
    levy_area_coeffs: AbstractLevyAreaCoeffs
    ...

class GeneralBrownianCoeffs(AbstractBrownianCoeffs):
    levy_area_coeffs: AbstractLevyAreaCoeffs
    ...
```
In other words, you can treat the additive-vs-general "axis" separately to the none-vs-spacetime-vs-spacetimetime "axis". Right now you have them flattened down into  just a single "axis".

(I'm using "axis" as you can maybe see the analogy between this and the axes of an array: mathematically it's a choice between the product set `{a, b} x {c, d, e}` and the single set `{ac, ad, ae, bc, bd, be}`.)

Reply:
Okay @patrick-kidger I thought about it, and I really don't think there's another way to do it more cleanly than the way it is now. The reason for this is that the type of a and b depends on whether they are inside `AdditiveCoeffs` or `GeneralCoeffs`, so in your example `AbstractLevyAreaCoeffs` would need to be parametrized by the type of a and b, which again makes illegal cases representable (because you could mix and match...) and makes the code long and confusing. Also making 5 instead of 6 classes (at the expense of much more complicated code) is probably not worth it.

Honestly, I'm not sure I'll believe a cleaner solution exists unless you show me the whole `srk.py` with it.